libsolidity/inlineasm/AsmData.h

```diff
@@ -58,23 +58,25 @@ struct StackAssignment { SourceLocation location; Identifier variableName; };
 /// Multiple assignment ("x, y := f()"), where the left hand side variables each occupy
 /// a single stack slot and expects a single expression on the right hand returning
 /// the same amount of items as the number of variables.
-struct Assignment { SourceLocation location; std::vector<Identifier> variableNames; std::shared_ptr<Statement> value; };
+struct Assignment { SourceLocation location; std::vector<Identifier> variableNames; std::shared_ptr<Expression> value; };
 /// Functional instruction, e.g. "mul(mload(20:u256), add(2:u256, x))"
-struct FunctionalInstruction { SourceLocation location; solidity::Instruction instruction; std::vector<Statement> arguments; };
-struct FunctionCall { SourceLocation location; Identifier functionName; std::vector<Statement> arguments; };
+struct FunctionalInstruction { SourceLocation location; solidity::Instruction instruction; std::vector<Expression> arguments; };
+struct FunctionCall { SourceLocation location; Identifier functionName; std::vector<Expression> arguments; };
+/// Statement that contains only a single expression
+struct ExpressionStatement { SourceLocation location; Expression expression; };
```

Comment:
Why does it need its own source location, the expression has one already? Is there a way to avoid this extra struct and use `boost:variant`?

Reply:
I'm not sure whether `boost::variant` here would work, it might create some problems in the templates, but let me try.

Reply:
Hm, I think it is cleaner to do the separation with another indirection here. `Expression` is the category of all expressions, `Statement` is the category of all statements and the `ExpressionStatement` is a statement that _contains_ an expression (and not _is_ an expression).

Reply:
The location is required because of the template function `locationOf` - all Statements need a location.